### PR TITLE
Remove mentions of the "Path Identifier"

### LINKED ISSUE #169: Does the Path ID change if the CID changes or not?
Currently it seems that the assumption is that the Path ID changes if the CID changes, however, this makes handling of MP frames that contain the path ID more complicated because you have to remember the old path ID to process these frames even when the CID is retired. Is there any good reason why we don't keep the Path ID constant over the life time of a path?

### LINKED ISSUE #181: Regular ACK frames when multipath has been negotiated
# Introduction

It should be allowed to send regular ACK frames also after multipath has been negotiated.

# Reasoning

Sending acknowledgments over other paths than where the packet was received is complex, which is recognised in the Packet Scheduling section of the draft. A straight-forward implementation may therefore never send cross-path acknowledgments. Allowing it to use regular ACK frames would reduce the special-case code for multipath mode and also save at least one byte on every acknowledgment.

The specification should be updated to (explicitly) say that an endpoint MAY send regular ACK frames also after multipath has been negotiated.

### PR DESCRIPTION
**Pull request by qdeconinck**
In lot of places in the draft, we mention the "Path Identifier", which we define as the Destination Connection ID sequence number used over that path. It seems that maintaining the notion of "Path Identifier" causes a lot of confusion, so better keep things simple and mention CID sequence numbers directly.

Note that this implies that the related fields in
multipath-specific have been renamed to their "explicit" form. Some of these names are maybe a bit long, so we may shorten them if this does not introduce ambiguity.

Fix #169. Rewriting related section may also answer #181.

### CODE REVIEW

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:212]:
```suggestion
A new path can only be used once it has been validated. The Destination
Connection ID is used to associate a packet to a valid path. Further, the
sequence number of Destination Connection ID is used as numerical identifier
in control frames. E.g. an endpoint sends a PATH_ABANDON frame to request its peer to
abandon the path on which the sender uses the Destination Connection ID
with the sequence number contained in the PATH_ABANDON frame.
```

Comment (mirjak) [reply draft-ietf-quic-multipath.md:212]: Maybe we can also move the text about the status frame here.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:536]: I don't think we need this additional text.

You don't need the sequence number to identify the packet space; you can also just use the CID directly. You only need the sequence number really in control frames.

Also what consideration in RFC9000 do you want to point to?

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:543]:
```suggestion
packets over the path. The endpoint relies on its sequence number to
send oath control information and specifically acknowledge packets belonging to that Connection ID-specific
packet number space.
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:591]:
```suggestion
Connection ID Sequence Number field. If multipath has been successfully
negotiated, ACK frames in 1-RTT packets MUST be interpreted as acknowledging
packets on the path they the packet with the ACK frames been received.
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:599]:
```suggestion
ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be sent on same the path
as the data was sent that it acknowledged.
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:964]:
```suggestion
  corresponds to.
```

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:543]: Typo in suggested change: oath -> path

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:591]: No, please do not apply this change. This PR is editorial in nature, and should not change the consensus so far, i.e., the old ACK frame acknowledges CID number 0. Quentin's original text is fine.

File: draft-ietf-quic-multipath.md
Comment (huitema) [inline draft-ietf-quic-multipath.md:1030]: I think we have an issue of consistency. In the Abandon and Path Status frames, the DCID field is called "Sender's DCID". Here, it is called "Receiver's DCID". Yet, they refer to exactly the same value. My suggestion:

* Remove "Sender's" from the DCID descriptions in PATH_ABANDON and PATH_STATUS
* Remove "Receiver's" from the DCID descriptions in ACK_MP
* Otherwise, keep the text as is, because the text is correct.

File: draft-ietf-quic-multipath.md
Comment (huitema) [inline draft-ietf-quic-multipath.md:1039]: The reference to "closed paths" is a bit ambiguous, because it is in fact OK to send an ACK_MP after and ABANDON_PATH -- or even in the same packet, but QUIC does not specify the order of processing of frames in a packet. And it is also OK to process it, rather than spuriously repeat data that has been actually received. I would be tempted to just strike out "or belonging to closed paths". Or maybe strike out the whole parenthesis. Or maybe strike out the whole paragraph about not active anymore. For comparison, Section 19.16 of RFC 9000 just says:

Receipt of a RETIRE_CONNECTION_ID frame containing a sequence number greater than any previously sent to the peer MUST be treated as a connection error of type PROTOCOL_VIOLATION.

**Review by huitema — COMMENTED** [review]
A few nits, but this PR does improve the draft!

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:543]: fixed

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:536]: Indeed, let's remove that added text.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:591]: Yes, this changes goes maybe a bit far for this PR but Quentin's text is not right because he removed the "If for any reason  ACK frames are received in 1-RTT packets while the state of multipath negotiation is ambiguous" part. In this ambiguous state all acks are mapped to the seq number 0. However, this is not true later. My proposal is not to change the consensus but rather say the more general thing.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:1030]: No they do not refer to the same value. The Sender's CID is the one used when you sent a packet. The Receivers's CID is the one used by the other end to send - hence when you receive packets.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:591]: Yes, let's avoid introducing design changes here. Also, I think relying on the reception 4-tuple may lead to ambiguous state at the receiver.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:599]: While I understand the suggestion, I'm not sure this makes the text clearer, though I don't have strong opposition here.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:1030]: Actually the PATH_ABANDON and PATH_STATUS frame includes the sequence number of the DCID of the sender of the frame, while in ACK_MP the sequence number corresponds to the DCID used by its peer, hence the (verbose) name convention. But if this introduces confusion, I can remove "Sender's" and "Receiver's" mentions.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:1034]: I think your point about the whole parenthesis is right. Also, this issue about seeing retired DCID also applies in PATH_ABANDON and PATH_STATUS (i.e., just ignore the frame), so we may just want to add a sentence at the beginning of the presentation of the multi path-specific frames?

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:586]: For the ambiguous part, I'm not sure to understand when the negotiation of multi path could be ambiguous. But so far the less ambiguous approach (as I understood so far) is that ACK frames (always) relate to CID number 0.

Comment (mirjak) [reply draft-ietf-quic-multipath.md:586]: I was assuming an ACK frame always corresponds to the path it was sent on. However, it seem this is simply not defined in the draft and we probably need to open an issue to clarify this.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:1025]: Actually we did a mistake here when I remove the type field. The path abandon frame should also have the receivers cid as you tell the other end to not use this for sending data anymore. Same I probably true for the path status frame.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:586]: No, we specifically do not want implementations to continue sending ACK frames after negotiating multipath. The notion of "current path" is ambiguous -- for example, it does change in case of NAT rebinding. Adding the CID-sequence-number in the ACK_MP guarantee that there is no such ambiguity. As for nake ACK, the equivalence to "cid-number=0" is straightforward.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:1025]: I think aligning ACK, ABANDON and STATUS makes sense. ABANDON is in many respects just the symmetric of the ACK. For example, a node receives data on a cellular path that charges per byte when a "free" path is available. The node will send the ABANDON immediately, telling its peer to "stop doing that." Using the DCID "used by the peer to send 1-RTT packets" makes a lot of sense.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:594]: I fully agree here. Sending the ACK_MP on the same path is a reasonable default behaviour, but in some network configurations, this is clearly suboptimal and there is no point forcing endpoints to do that. I think having a  "SHOULD" is fine, but not a "MUST".

Rethinking further, I'm wondering if this paragraph could even be removed based on discussions in #183. We already discuss the ACK_MP scheduling in implementation considerations, and that should be enough. Maybe add at some place that frames can effectively be sent and received on any valid/active path.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:1025]: Yeah, this would make sense. Let me update the text and the examples accordingly.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:208]:
```suggestion
A new path can only be used once the associated 4-tuple has been validated
by ensuring that the peer is able to receive packets at that address
(see {{Section 8 of RFC9000}}). The Destination Connection ID is used
to associate a packet to a packet number space that is used on a valid path. Further, the
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:377]:
```suggestion
which request the peer to stop sending packets with the corresponding Destination Connection ID.
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:375]:
```suggestion
Both endpoints, namely the client and the server, can initiate path closure,
```

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:185]: While the path actually refers to the (perceived) 4-tuple, not sure we say the path only depends on the CID. The CID (sequence number) is used between endpoints to identify a path in PATH_ABANDON/PATH_STATUS frames. However, we can remove this definition if this brings confusion.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:584]:
```suggestion
Sequence Number field to identify the packet number space.
If the multipath extension has been successfully
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:594]:
```suggestion
ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be sent on the path
it received packet with the Connection ID of the packet number space it acknowledges.
```

Comment (mirjak) [reply draft-ietf-quic-multipath.md:594]: On the editorial part: I think "perceives" is just the wrong word. I was also trying to make it shorter but feel free to take any of the two suggestions.

On the text itself, if we use normative language it should stay here and not move to the implementation guidance section. let keep this for now and maybe open a separate issue in case we need more discussion or actually want to change it.

My view is: I think we need some normative guidance because of the RTT calculation issue. However, there might be room for refining it or provide further explanation in the implementation guidance scheduling section.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:1031]: I see that this text was remove, which is fine. But was it re-added somewhere else?

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: "receive packet with" is a better text than "perceived".

And yes, let's merge this PR as is, and discuss path affinity of MP_ACK separately.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:378]:
```suggestion
by sending a PATH_ABANDON frame (see {{path-abandon-frame}}) which
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:381]: This is entirely my fault with PR #174 but this is actually specified already later in the text and therefore redundant here. However, we can also just leave this for now and make another editorial pass later.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: I started a discussion of the "ACK on same path" in issue #190

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:1034]: Not yet. We may add a sentence to clarify that we should just ignore Multipath frames related to retired DCIDs, but I would suggest to add such a sentence at the beginning of the section introducing these frames, as this applies to all of them.

Comment (mirjak) [reply draft-ietf-quic-multipath.md:1029]: Okay, maybe let's open an issue, so we don't forget!

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:376]: Okay, should we open an (editorial) issue, so we don't forget? Or we can create a new PR right after this one is merged (without an issue).

**Comment by mirjak** [comment]
I think this PR is ready to merge. Let's wait until end of today to give others another change to review. However would be good to merge to, as we need to merge this one before we can merge the others!

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:1029]: Filled #193.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:376]: Filled #194.

**Review by Yanmei-Liu — APPROVED** [review]
Looks good to me.

**Pull request merged**
